File: client/speechly.js
require("draftlog").into(console);

const fs = require("fs");
const grpc = require("grpc");
const protoLoader = require("@grpc/proto-loader");
const prompts = require("prompts");
const uuid = require("uuid");

exports.getSpeechlyClient = async () => {
  const keys = {};
  const keyfile = ".speechly_token";
  if (fs.existsSync(keyfile)) {
    try {
      const storedKeys = JSON.parse(fs.readFileSync(keyfile));
      Object.assign(keys, storedKeys);
      if (new Date() - new Date(keys.issued) > 1000 * 60) {
        console.log("token is too old");
        keys.issued = undefined;
      }
    } catch (err) {
      console.error("Invalid token file, recreating");
    }
  }
  if (keys.issued === undefined) {
    if (keys.appId === undefined) {
      const response = await prompts([
        { name: "appId", type: "text", message: "appId" },
        { name: "languageCode", type: "text", message: "languageCode", initial: "en-US" }
      ]);
      keys.appId = response.appId;
      keys.languageCode = response.languageCode;
    }
    if (keys.deviceId === undefined) {
      keys.deviceId = uuid.v4();
    }
    const client = createSpeechlyClient();
    keys.token = await new Promise((resolve, reject) => {
      client.identity.Login({ appId: keys.appId, deviceId: keys.deviceId }, (err, response) => {
        if (err) {
          return reject(err);
        }
        console.log(`Logged in device ${keys.deviceId} to application ${keys.appId}`);
        return resolve(response.token);
      });
    });
    keys.issued = new Date().toISOString();
    fs.writeFileSync(keyfile, JSON.stringify(keys, undefined, 2));
  }
  return createSpeechlyClient(keys.token, keys.languageCode);
};

const createSpeechlyClient = (token, languageCode) => {
  let host = "api.speechly.com";
  let creds = grpc.credentials.createSsl();
  if (process.env["URL"] !== undefined) {
    host = process.env["URL"];
    creds = grpc.credentials.createInsecure();
  }
  const Speechly = grpc.loadPackageDefinition(
    protoLoader.loadSync("../speechly.proto", {
      keepCase: false,
      longs: String,
      enums: String,
      defaults: true,
      oneofs: true
    })
  );

  let metadata = undefined;
  if (token !== undefined) {
    //const speechlyCreds = grpc.credentials.createFromMetadataGenerator((args, callback) => {
    metadata = new grpc.Metadata();
    metadata.add("Authorization", `Bearer ${token}`);
    //   callback(null, metadata);
    // });
    // creds = grpc.credentials.combineChannelCredentials(creds, speechlyCreds);
  }

  return {
    languageCode,
    metadata,
    slu: new Speechly.v1.SLU(host, creds),
    wlu: new Speechly.v1.WLU(host, creds),
    identity: new Speechly.v1.Identity(host, creds)
  };
};

File: python/speechly.py
import datetime
import json
import os
import uuid

import grpc

from speechly_pb2 import LoginRequest, WLURequest
from speechly_pb2_grpc import IdentityStub, WLUStub


KEY_FILE = '.speechly_token'

class Speechly:
    def __init__(self, app_id: str):
        channel = grpc.secure_channel("api.speechly.com", grpc.ssl_channel_credentials())
        self.identity_stub = IdentityStub(channel)
        self.wlu_stub = WLUStub(channel)
        self.metadata = []
        self._login(app_id)

    def _login(self, app_id:str):
        keys = {}
        if os.path.exists(KEY_FILE):
            with open(KEY_FILE, 'r') as fp:
                keys = json.load(fp)
        if keys.get('issued') is None:
            keys['appId'] = app_id
            if keys.get('deviceId') is None:
                keys['deviceId'] = str(uuid.uuid4())
            login_request = LoginRequest(device_id=keys['deviceId'], app_id=keys['appId'])
            response = self.identity_stub.Login(login_request)
            if not response.token:
                raise Exception('No token in response')
            keys['token'] = response.token
            keys['issued'] = datetime.datetime.now(datetime.timezone.utc).isoformat()
            with open(KEY_FILE, 'w') as fp:
                json.dump(keys, fp)
        self.metadata = [('authorization', 'Bearer ' + keys['token'])]

    def wlu(self, text: str, language_code: str = 'en-US'):
        request = WLURequest(language_code=language_code, text=text)
        return self.wlu_stub.Text(request, metadata=self.metadata)
